pkg/webhook/workspace/handler/metadata.go

```diff
@@ -0,0 +1,101 @@
+//
+// Copyright (c) 2019-2020 Red Hat, Inc.
+// This program and the accompanying materials are made
+// available under the terms of the Eclipse Public License 2.0
+// which is available at https://www.eclipse.org/legal/epl-2.0/
+//
+// SPDX-License-Identifier: EPL-2.0
+//
+// Contributors:
+//   Red Hat, Inc. - initial API and implementation
+package handler
+
+import (
+	"fmt"
+	"github.com/che-incubator/che-workspace-operator/pkg/controller/workspace/model"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func (h *WebhookHandler) mutateMetadataOnCreate(o *metav1.ObjectMeta) error {
+	if o.GetAnnotations() == nil {
+		return fmt.Errorf("annotations are missing while %s is required", model.WorkspaceCreatorAnnotation)
+	} else {
+		if _, ok := o.GetAnnotations()[model.WorkspaceCreatorAnnotation]; !ok {
+			return fmt.Errorf("'%s' annotation is missing", model.WorkspaceCreatorAnnotation)
+		}
+	}
+
+	if o.GetLabels() == nil {
+		return fmt.Errorf("labels are missing while %s is required", model.WorkspaceIDLabel)
+	} else {
+		if _, ok := o.GetLabels()[model.WorkspaceIDLabel]; !ok {
+			return fmt.Errorf("'%s' label is missing", model.WorkspaceIDLabel)
+		}
+	}
+
+	return nil
+}
+
+func (h *WebhookHandler) mutateMetadataOnUpdate(oldMeta, newMeta *metav1.ObjectMeta) (bool, error) {
+	if oldMeta.Annotations == nil {
+		oldMeta.Annotations = map[string]string{}
```

Comment:
Isn't `oldMeta.Annotations == nil` or `oldMeta.Labels == nil` an automatic failure? It's strange to create maps here only to check for a specific value in them later.

Alternatively, we could fallback to nil map ~= empty map when reading here.

Reply:
You're right, I missed that nil map in go behaves as an empty map in case of reading. So, it makes sense to initialize only newLabels, and newAnnotations since they may be modified.